hudi-common/src/main/java/org/apache/hudi/metadata/FileSystemBackedTableMetadata.java

```diff
@@ -49,12 +60,48 @@ public FileSystemBackedTableMetadata(SerializableConfiguration conf, String data
 
   @Override
   public List<String> getAllPartitionPaths() throws IOException {
-    FileSystem fs = new Path(datasetBasePath).getFileSystem(hadoopConf.get());
     if (assumeDatePartitioning) {
+      FileSystem fs = new Path(datasetBasePath).getFileSystem(hadoopConf.get());
       return FSUtils.getAllPartitionFoldersThreeLevelsDown(fs, datasetBasePath);
-    } else {
-      return FSUtils.getAllFoldersWithPartitionMetaFile(fs, datasetBasePath);
     }
+
+    List<Path> pathsToList = new LinkedList<>();
```

Comment:
is this block verbatim the code @prashantwason had for parallelizing bootstrapping.. Prashant, can you also please help review this PR

Reply:
Yeah I adapted the same code with some changes which optimize for this particular case of fetching only partition paths.

Reply:
Can we add more tests to this piece of code? (previously it was just in the bootstrap part), not this is prime time.

Reply:
As I had mentioned, that this is already tested through code paths like cleaner, and I did verify that `TestCleaner` for example fails if this would return incorrect partitions.

Adding specific tests for `FileSystemBackedMetadata` is tricky because `spark context` is not available in `hudi-common`. Perhaps, this can be done in a separate PR.